src/pkg/shub/client/util.go

```diff
@@ -0,0 +1,80 @@
+// Copyright (c) 2018, Sylabs Inc. All rights reserved.
+// This software is licensed under a 3-clause BSD license. Please consult the
+// LICENSE.md file distributed with the sources of this project regarding your
+// rights to use or distribute this software.
+
+package client
+
+import (
+	"fmt"
+	"regexp"
+	"strings"
+)
+
+// ShubParseReference accepts a Shub reference string and parses its content
+// It will return an error if the given URI is not valid,
+// otherwise it will parse the contents into a ShubURI struct
+func ShubParseReference(src string) (uri ShubURI, err error) {
+
+	//define regex for each URI component
+	registryRegexp := `([-a-zA-Z0-9/]{1,64}\/)?` //target is very open, outside registry
+	nameRegexp := `([-a-zA-Z0-9]{1,39}\/)`       //target valid github usernames
+	containerRegexp := `([-_.a-zA-Z0-9]{1,64})`  //target valid github repo names
+	tagRegexp := `(:[-_.a-zA-Z0-9]{1,64})?`      //target is very open, file extensions or branch names
+	digestRegexp := `(\@[a-f0-9]{32})?`          //target md5 sum hash
```

Comment:
After an '@' may not be an md5 hash:

https://github.com/singularityhub/singularityhub.github.io/wiki/Deploy#order-of-operations

> shub://vsoch/hello-world:@ is where it get's interesting. A digest is flexible to being the following, in the order of preference:
> hash the most specific kind of digest is the hash of the file. This is checked for first.
> commit the next most specific is the commit
> branch finally, if a matching hash and commit are not found, we fall back to the branch.

Reply:
you mean just to adjust the comment right?

Reply:
I mean currently this doesn't support all shub URIs... you can only pull xxxx@<hash> not xxx@<commit> or xxxx@<branch>